crates/astria-core/src/primitive/v1/mod.rs

```diff
@@ -490,46 +491,83 @@ where
     tree
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 #[cfg_attr(
     feature = "serde",
     serde(try_from = "raw::TransactionId", into = "raw::TransactionId")
 )]
 pub struct TransactionId {
-    hash: String,
+    inner: [u8; TRANSACTION_ID_LEN],
 }
 
 impl TransactionId {
-    /// Returns the hex-encoded 32-byte transaction hash as a string.
+    /// Constructs a new `TransactionId` from a 32-byte array.
     #[must_use]
-    pub fn get(&self) -> String {
-        self.hash.clone()
+    pub const fn new(inner: [u8; TRANSACTION_ID_LEN]) -> Self {
+        Self {
+            inner,
+        }
+    }
+
+    /// Returns the 32-byte transaction hash.
+    #[must_use]
+    pub fn get(self) -> [u8; TRANSACTION_ID_LEN] {
+        self.inner
     }
 
     #[must_use]
     pub fn to_raw(&self) -> raw::TransactionId {
         raw::TransactionId {
-            hash: self.hash.clone(),
+            inner: hex::encode(self.inner),
         }
     }
 
     #[must_use]
     pub fn into_raw(self) -> raw::TransactionId {
         raw::TransactionId {
-            hash: self.hash.clone(),
+            inner: hex::encode(self.inner),
         }
     }
 
-    /// Returns a copy of the transaction hash as a 32-byte array.
+    /// Convert from a raw protobuf type to a rust type for a transaction ID.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the transaction ID buffer was not 32 bytes long or if it was not hex
+    /// encoded.
+    pub fn try_from_raw(raw: raw::TransactionId) -> Result<Self, TransactionIdError> {
```

Comment:
You can implement this in terms of `try_from_raw_ref` because we don't need to clone anything. Less code duplication.